Source/santa-driver/SantaPrefixTree.cc

```diff
@@ -0,0 +1,238 @@
+      // If the upgrade fails, the shared lock is released.
+      if (!lck_rw_lock_shared_to_exclusive(spt_lock_)) {
+        // Grab a new exclusive lock.
+        lck_rw_lock_exclusive(spt_lock_);
+      }
+
+      // Is there enough room for the rest of the prefix?
+      if (!branched && (node_count_ + (len - i)) > kMaxNodes) {
+        LOGE("Prefix tree is full, can not add: %s", prefix);
+        if (node_count) *node_count = node_count_;
+        lck_rw_unlock_exclusive(spt_lock_);
+        return kIOReturnNoResources;
+      }
+
+      SantaPrefixNode *new_node = new SantaPrefixNode();
+
+      // If this is the end, mark the node as a prefix.
+      if (i + 1 == len) {
+        LOGD("Added prefix: %s", prefix);
+        new_node->isPrefix = true;
+      }
+
+      node->children[value] = (void *)new_node;
+      ++node_count_;
+      branched = true;
+
+      // Downgrade the exclusive lock
+      lck_rw_lock_exclusive_to_shared(spt_lock_);
+    } else if (i + 1 == len) {
+      // If the child does exist and it is the end...
+      // Set the new, higher prefix and prune the now dead nodes.
+
+      if (!lck_rw_lock_shared_to_exclusive(spt_lock_)) {
+        lck_rw_lock_exclusive(spt_lock_);
+      }
+
+      PruneNode((SantaPrefixNode *)node->children[value]);
+
+      SantaPrefixNode *new_node = new SantaPrefixNode();
+      new_node->isPrefix = true;
+
+      node->children[value] = (void *)new_node;
+      ++node_count_;
+
+      LOGD("Added prefix: %s", prefix);
+
+      lck_rw_lock_exclusive_to_shared(spt_lock_);
+    }
+
+    // Get ready for the next iteration.
+    node = (SantaPrefixNode *)node->children[value];
+  }
+
+  if (node_count) *node_count = node_count_;
+
+  lck_rw_unlock_shared(spt_lock_);
+
+  return kIOReturnSuccess;
+}
+
+bool SantaPrefixTree::HasPrefix(const char *string) {
+  lck_rw_lock_shared(spt_lock_);
+
+  auto found = false;
+
+  SantaPrefixNode *node = root_;
+
+  // TODO(bur): Avoid using strlen().
+  size_t len = strlen(string);
+  for (int i = 0; i < len; ++i) {
+    // Only process a byte at a time.
+    uint8_t value = string[i];
+
+    // Get the next char string.
+    node = (SantaPrefixNode *)node->children[value];
+
+    // If it doesn't exist in the tree, no match.
+    if (!node) break;
+
+    // If it does exist, is it a prefix?
+    if (node->isPrefix) {
+      found = true;
+      break;
+    }
+  }
+
+  lck_rw_unlock_shared(spt_lock_);
+
+  return found;
+}
+
+void SantaPrefixTree::Reset(uint32_t *node_count) {
+  lck_rw_lock_exclusive(spt_lock_);
+
+  PruneNode(root_);
+  root_ = new SantaPrefixNode();
+  ++node_count_;
+
+  if (node_count) *node_count = node_count_;
+  LOGD("Prefix tree reset");
+
+  lck_rw_unlock_exclusive(spt_lock_);
+}
+
+void SantaPrefixTree::PruneNode(SantaPrefixNode *target) {
+  if (!target) return;
+
+  // For deep trees a recursive approach will generate too many stack frames.
+  auto stack = OSArray::withCapacity(1);
+  auto seen = OSSet::withCapacity(1);
```

Comment:
Start with the capacity of node_count_ - you might not need all that space as you might be starting part way down the tree but better that then to keep having to re-double the size.